### PR TITLE
Feat java orb

### PR DESCRIPTION
**Pull request by kazi-ricardo**
~POC for java orb based on selling-domain-api~

to minimize configuration in java repos we introduce
- new java specific `maven_build_test` job
- new java specific `maven_*` commands
- new java specific `maven_*` executors

based on and tested with proof of concept configuration in `incidents-api`, see https://github.com/ricardo-ch/incidents-api/pull/172

‼️ contains some (small-ish) breaking changes from #20

### CODE REVIEW

File: src/jobs/java_isopod_deploy.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  slack_fail_branches:
+    description: "Comma separated list of branches for which notifications are sent"
+    type: string
+    default: "master"
+
+executor:
+  name: isopod
+  tag: << parameters.isopod_version >>
+  username: << parameters.artifactory_username >>
+  password: << parameters.artifactory_password >>
```
Comment (michaelkuechler) [inline src/jobs/java_isopod_deploy.yml:59]: @robert-nemet do you think this pseudo-polymorph executor can work?

if i validate the orb with `circleci orb pack --skip-update-check src > orb.yml ; circleci orb --skip-update-check validate orb.yml` i notice that

if `java_builder_docker` does not specify the `vm_image` parameter, validation fails with
```
Error: Error calling job: 'java_maven_build_test'
Error calling executor: 'java_builder_docker'
Unexpected argument(s): vm_image
```
however if i add the `vm_image` parameter there (dummy param, not used), validation does _NOT_ fail for `java_builder_vm`, which i would expect to have the same problem with the `java_version` parameter that is only present on `java_builder_docker`

TLDR: im a bit afraid that the resulting orb does not work for the `java_builder_vm`, and that the validation does not find the problem because it does not dynamically resolve the executor name and rather bakes something in the first time it happens to see a value 🤔

File: src/jobs/java_maven_build_test.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    default: "11.0"
+  java_builder_vm_image:
+    description: "[required for java_builder_vm] VM-image for executor"
+    type: string
+    default: "ubuntu-2004:202010-01"
+
+executor:
+  name: << parameters.executor >>
+  java_version: << parameters.java_builder_docker_java_version >>
+  vm_image: << parameters.java_builder_vm_image >>
```
Comment (michaelkuechler) [inline src/jobs/java_maven_build_test.yml:31]: @robert-nemet do you think this pseudo-polymorph executor can work?

if i validate the orb with `circleci orb pack --skip-update-check src > orb.yml ; circleci orb --skip-update-check validate orb.yml` i notice that

if `java_builder_docker` does not specify the `vm_image` parameter, validation fails with
```
Error: Error calling job: 'java_maven_build_test'
Error calling executor: 'java_builder_docker'
Unexpected argument(s): vm_image
```
however if i add the `vm_image` parameter there (dummy param, not used), validation does _NOT_ fail for `java_builder_vm`, which i would expect to have the same problem with the `java_version` parameter that is only present on `java_builder_docker`

TLDR: im a bit afraid that the resulting orb does not work for the `java_builder_vm`, and that the validation does not find the problem because it does not dynamically resolve the executor name and rather bakes something in the first time it happens to see a value 🤔

File: src/commands/java_isopod_build.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      condition: << parameters.appname >>
+      steps:
+        - run: |
+            isopod -f << parameters.appname >>/isopod.yml build
+  # build app image from single-app-repo
+  - unless:
+      condition: << parameters.appname >>
+      steps:
+        - run: |
+            isopod build
```
Comment (robert-nemet) [inline src/commands/java_isopod_build.yml:23]: These two are the same:
```shell
isopod build -f ./isopod.yml
```

and:

```shell
isopod build
```
Taking this in account `appname` default value could be `.`.

File: src/commands/java_cache_maven_artifacts.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+description: >
+  Cache maven artifacts.
+  Works with monorepo as well as single-app-repo.
+
+parameters:
+  maven_cache_key_prefix:
+    description: "Prefix for the cache-key (used in combination with checksum of pom.xml)"
+    type: string
```
Comment (robert-nemet) [inline src/commands/java_cache_maven_artifacts.yml:8]: Maybe we could have default value? Application name maybe?

File: src/commands/java_isopod_deploy.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+description: >
+  Deploy a java app.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+  Expects an isopod.yml file to be present in the app directory (maven module directory for monorepo app, or project root for single-app-repo).
+
+parameters:
+  appname:
+    description: "Name of maven module for the app, or blank for single-app-repo"
+    type: string
+    default: ""
```
Comment (robert-nemet) [inline src/commands/java_isopod_deploy.yml:10]: If default is `.` (current directory) maybe we do not need `unless` block?

File: src/commands/java_restore_maven_artifacts.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+description: >
+  Restore cached maven artifacts.
+  Works with monorepo as well as single-app-repo.
+
+parameters:
+  maven_cache_key_prefix:
+    description: "Prefix for the cache-key (used in combination with checksum of pom.xml); no-op with blank prefix"
+    type: string
+    default: ""
```
Comment (robert-nemet) [inline src/commands/java_restore_maven_artifacts.yml:9]: I wonder if prefix default value can be app name?

File: src/commands/java_save_maven_output.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+description: >
+  Save maven output directory (target directory containing compiled app) to workspace.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+
+parameters:
+  appname:
+    description: "Name of maven module for the app, or blank for single-app-repo"
+    type: string
+    default: ""
```
Comment (robert-nemet) [inline src/commands/java_save_maven_output.yml:9]: Maybe default value can be `.`?  If yes, then we can remove `unless` block

File: src/commands/java_isopod_deploy.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+description: >
+  Deploy a java app.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+  Expects an isopod.yml file to be present in the app directory (maven module directory for monorepo app, or project root for single-app-repo).
```
Comment (robert-nemet) [inline src/commands/java_isopod_deploy.yml:4]: Why this would be Java specific? For deploying only `isopod.yaml` and branch is important. It is stack agnostic

File: src/commands/print.yml [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+parameters:
+  message:
+    description: "Message to print to console"
+    type: string
+    default: "Empty message"
+
+steps:
+  - run: |
+      echo "<< parameters.message >>"
```
Comment (robert-nemet) [inline src/commands/print.yml:12]: Why?

File: src/executors/java_builder_docker.yml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  java_version:
+    description: "Java version to install (or rather selecting an existing image accordingly)"
+    type: string
+    default: "11.0"
+  docker_hub_username:
+    type: string
+    default: $DOCKER_JFROG_USERNAME
+  docker_hub_password:
+    type: string
+    default: $DOCKER_JFROG_PASSWORD
```
Comment (robert-nemet) [inline src/executors/java_builder_docker.yml:14]: `cimg/openjdk` is stored on docker hub, not on JFrog, so you need docker credentials.

File: src/jobs/java_isopod_build_push_image.yml [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  name: isopod
+  tag: << parameters.isopod_version >>
+  username: << parameters.private_hub_username >>
+  password: << parameters.private_hub_password >>
+
+steps:
+  - checkout
+  - maven_auth
+  - setup_remote_docker:
+      docker_layer_caching: false
```
Comment (robert-nemet) [inline src/jobs/java_isopod_build_push_image.yml:44]: Why false?

File: src/jobs/java_isopod_deploy.yml [hunk 12]
```diff
@@ -0,0 +1,69 @@
+description: >
+  Deploy java app image to environment using isopod.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+
```
Comment (robert-nemet) [inline src/jobs/java_isopod_deploy.yml:4]: For deploying apps to the platform I do not see reason to have stack specific jobs/commands unless they are in separate orbs. Reason is that isopod do not care about stacks..

File: src/jobs/java_maven_build_test.yml (hunk 2 above)
Comment (michaelkuechler) [reply src/jobs/java_maven_build_test.yml:31]: > Executor `java_builder_docker` do not support parameter `vm_image`.

yes i know, i wrote it 😂 
im just wondering why the other way around the problem is not found in validation - i.e. if i add the `vm_image` param to `java_builder_docker`, the validation does not realize that there is params from `java_builder_docker` that are not supported on `java_builder_vm` 😟

File: src/commands/java_restore_maven_artifacts.yml (hunk 6 above)
Comment (michaelkuechler) [reply src/commands/java_restore_maven_artifacts.yml:9]: see [above](https://github.com/ricardo-ch/ricardo-orbs/pull/17#discussion_r914538969) - IMHO it does not make sense

File: src/commands/print.yml (hunk 9 above)
Comment (michaelkuechler) [reply src/commands/print.yml:12]: true - can be inlined 🤦‍♂️ 
i was copy pasting too much here

File: src/commands/java_isopod_deploy.yml (hunk 8 above)
Comment (michaelkuechler) [reply src/commands/java_isopod_deploy.yml:4]: i guess it can be inlined into the java_isopod_deploy job

File: src/jobs/java_isopod_build_push_image.yml (hunk 11 above)
Comment (michaelkuechler) [reply src/jobs/java_isopod_build_push_image.yml:44]: this is inline+customize of existing job [build_push_image](https://github.com/ricardo-ch/ricardo-orbs/blob/feat-java-orb/src/jobs/build_push_image.yml#L76-L78) - and there this is the default

since there was no other configuration in incidents-api config i went with this
should we need to make this configurable, i will extract this into a param with default=false (like in build_push_image job)

File: src/jobs/java_isopod_deploy.yml (hunk 12 above)
Comment (michaelkuechler) [reply src/jobs/java_isopod_deploy.yml:4]: isopod not caring about stacks i agree

me having to use a job that is build to support node+java+go+brainfuck means i always have to fight through a billion optional customizations which i dont need (want) to know

the idea here is to have a job which requires minimal parameters for my java app config, sacrificing reusability for everything except my standard java app. even with that im still ending up with 300 lines that i have to copy+paste to each repo, which is mind-boggling!

File: src/jobs/java_maven_build_test.yml (hunk 2 above)
Comment (michaelkuechler) [reply src/jobs/java_maven_build_test.yml]: i refactored this now since i think it would not work and fail at runtime (even though validation did not catch the problem)

6b19880b81b568ef74ecaaa16ed972a30b3ec9e2 introduces a custom brute-forced polymorph behaviour that is closest to what i would like to have for an easy configuration in our java repos

**Review by robert-nemet — APPROVED** [review]

File: src/jobs/java_maven_build_test.yml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  java_builder_docker_hub_password:
+    description: "Docker hub credentials to download the executor image"
+    type: string
+    default: $DOCKER_HUB_PASSWORD
+
+executor:
+  name: << parameters.executor >>
+  image: << parameters.java_builder_image >>
+  docker_hub_username: << parameters.java_builder_docker_hub_username >>
+  docker_hub_password: << parameters.java_builder_docker_hub_password >>
```
Comment (julien-may) [inline src/jobs/java_maven_build_test.yml:36]: to me this seems like we always use docker and never a machine based executor. am i wrong?

File: src/jobs/java_isopod_deploy.yml (hunk 12 above)
Comment (julien-may) [reply src/jobs/java_isopod_deploy.yml:4]: i understand your comment @michaelkuechler but also agree with @robert-nemet . i propose we adapt the current `deploy_job` and add there are parameter to define where the isopod config file can be found. basically what you did here with `app_path`

File: src/commands/java_isopod_build.yml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+description: >
```
Comment (julien-may) [inline src/commands/java_isopod_build.yml:1]: just call id `isopod_build.yml`

File: src/commands/java_isopod_build.yml [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+  Build docker image for a java app using isopod, and push it to docker hub.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+
+parameters:
+  app_path:
```
Comment (julien-may) [inline src/commands/java_isopod_build.yml:6]: maybe just call it `path`?

File: src/commands/java_cache_maven_artifacts.yml [hunk 16]
```diff
@@ -0,0 +1,14 @@
+description: >
+  Cache maven artifacts.
+  Works with monorepo as well as single-app-repo.
+
+parameters:
+  maven_cache_key_prefix:
```
Comment (julien-may) [inline src/commands/java_cache_maven_artifacts.yml:6]: maybe just call it `prefix`?

File: src/commands/java_save_maven_output.yml [hunk 17]
```diff
@@ -0,0 +1,15 @@
+description: >
+  Save maven output directory (target directory containing compiled app) to workspace.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+
+parameters:
+  app_path:
```
Comment (julien-may) [inline src/commands/java_save_maven_output.yml:6]: maybe call it `module`?

File: src/jobs/java_maven_build_test.yml [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    default: ""
+  executor:
+    description: "Executor for the build (java_builder_docker is more lightweight, but some tests leveraging testcontainers etc require the java_builder_vm)"
+    type: enum
+    enum: [ "java_builder_docker", "java_builder_vm"]
+    default: "java_builder_docker"
+  java_builder_image:
+    description: "Image to use for the executor"
+    type: string
+    default: "cimg/openjdk:11.0"
```
[the remark this answers is not in the document]
Comment (michaelkuechler) [reply src/jobs/java_maven_build_test.yml:22]: that parameter makes the executor configurable in the java-application config

if i want to choose the executor by parameter and still be able to overwrite executor parameters (image version), the extension points of circleCI really dont provide what is needed. i cant even add a conditional when selecting/parametrizing the executor!

so what i did was
- select the executor by name (`java_builder_*` parameter)
- have a common set of params for all `java_builder_*` exectuors so that i can inject the required parameters/defaults without knowing where the values actually go. - a.k.a. polymorphism

if we dont need the power to overwrite the images in the java-repository configs, we can simplify here. but as long as we want to keep that capability there is no other way to do this AFAIK

File: src/commands/java_maven_build_test.yml [hunk 19]
```diff
@@ -0,0 +1,31 @@
+description: >
+  Execute a maven build including integration tests for a java app.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+
+parameters:
+  app_path:
+    description: "Path of maven module for the app, or \".\" for single-app-repo"
+    type: string
+    default: "."
```
[the remark this answers is not in the document]
Comment (michaelkuechler) [reply src/commands/java_maven_build_test.yml:9]: for monorepo you need to overwrite this param (we cant know the correct path otherwise). the default here is for the non-monorepo case (and allows to ignore the param in those configurations)

File: src/commands/java_maven_build_test.yml [hunk 20]
```diff
@@ -0,0 +1,31 @@
+description: >
+  Execute a maven build including integration tests for a java app.
+  Supports app-module from monorepo as well as single app from single-app-repo.
+
+parameters:
+  app_path:
```
[the remark this answers is not in the document]
Comment (michaelkuechler) [reply src/commands/java_maven_build_test.yml:6]: i was there, but if we think of it as the module path instead, the defaults become much nicer and suffice for the non-monorepo case.

File: src/jobs/java_maven_build_test.yml (hunk 13 above)
Comment (michaelkuechler) [reply src/jobs/java_maven_build_test.yml:36]: no. the executor is selected with the executor param (by name)

File: src/jobs/java_maven_build_test.yml [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+    type: string
+    default: "."
+  maven_cache_key_prefix:
+    description: "Prefix for the maven artifacts cache-key (used in combination with checksum of pom.xml); no caching with blank prefix"
+    type: string
+    default: ""
+  executor:
+    description: "Executor for the build (java_builder_docker is more lightweight, but some tests leveraging testcontainers etc require the java_builder_vm)"
+    type: enum
+    enum: [ "java_builder_docker", "java_builder_vm"]
```
[the remark this answers is not in the document]
Comment (michaelkuechler) [reply src/jobs/java_maven_build_test.yml]: docker is a reserved thing i guess
compromise: `maven_docker` and `maven_vm`

PS: subdirectories in src/executors dont work because the orb compiler takes that as orbname 🙄

File: src/jobs/java_maven_build_test.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (michaelkuechler) [reply src/jobs/java_maven_build_test.yml]: simplified now - by hardcoding the executor parametrization we can simplify a lot here
